### PR TITLE
Ipf

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
Adds in IPF functionality.

This has also become a working branch for landuse, so lots of general changes for that also.
Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

File: src/caf/core/data_structures.py [hunk 1]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+                    raise TranslationError(
+                        "For IPF any targets must either be at the same zoning "
+                        "system as the seed DVector, or be at a zoning system "
+                        "which the seed nests perfectly within. The zone_translation "
+                        "found contains non-one factors, which implies the "
+                        "zoning system doesn't nest, so IPF can't be performed."
+                    )
+        return targets
+
+    def ipf(self, targets: list[IpfTarget], tol: float = 1e-5, max_iters: int = 100):
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1354]: Use `collections.abc.Collection` instead of list in parameter type annotations because it is less restrictive. Also, please could you add a return type annotation to help clarify outputs for users.

File: src/caf/core/data_structures.py [hunk 2]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+        """
+        Implement iterative proportional fitting for DVectors.
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1356]: Is it worth adding a link to iterative propotional fitting methodology, or theory, to give more details.

File: src/caf/core/data_structures.py [hunk 3]
```diff
@@ -706,5 +1448,292 @@ def old_to_new_dvec(import_data: dict):
...
+            import_data=balanced,
+            segmentation=self.segmentation,
+            zoning_system=self.zoning_system,
+            time_format=self.time_format,
+        )
+
+
+@dataclass
+class IpfTarget:
+    """Dataclass to store targets to pass to IPF method of DVector."""
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1729]: Could the docstring explaing what the attributes are as I don't think it's completely obvious from the names. Also, could a name attribute be added to help with reporting it would potentially just be used to distinguish productions, attractions etc...

File: src/caf/core/data_structures.py [hunk 4]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+
+        Returns
+        -------
+        Input targets, with zone translations added in if relevant.
+        """
+        target_sum = 0
+        for position, target in enumerate(targets):
+            # Check targets sum to the same, or they can't converge. Potentially could allow
+            # IPF for non-agreeing targets to get as close as possible.
+            zoning_diff = False
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1290]: This is set twice but never used, should this be `target.zoning_diff`?

File: src/caf/core/data_structures.py [hunk 5]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+            if target_sum == 0:
+                target_sum = target.data.sum()
+            else:
+                # TODO don't hard code this
+                if not math.isclose(target_sum, target.data.sum(), abs_tol=target_sum / 1e5):
+                    raise ValueError(
+                        "Input target DVectors do not have consistent "
+                        "sums, so ipf will fail."
+                    )
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1299]: Error should detail which IPD target is has caused it, just adding position to the error is probably enough

File: src/caf/core/data_structures.py [hunk 6]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+            # Check segmentations are compatible.
+            if not target.data.segmentation.is_subset(self.segmentation):
+                if target.segment_translations is None:
+                    raise ValueError(
+                        "The target segmentation is not a subset of the seed "
+                        "segmentation, but no correspondences are defined."
+                    )
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1306]: Include position in error message to show which target caused the error

File: src/caf/core/data_structures.py [hunk 7]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+                                raise FileNotFoundError(
+                                    f"No segment translation found for {lower_seg} to {seg}.from."
+                                )
+
+            # Check zoning systems are compatible.
+            if self.zoning_system != target.data.zoning_system:
+                zoning_diff = True
+                if target.zone_translation is not None:
+                    pass
+                else:
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1328]: If nothing is happening in the if then this could be simplified
```suggestion
                if target.zone_translation is None:
```

File: src/caf/core/data_structures.py [hunk 8]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+                                # Don't need this for now, checking it exists.
+                                lower_seg.translate_segment(seg)
+                            except FileNotFoundError:
+                                raise FileNotFoundError(
+                                    f"No segment translation found for {lower_seg} to {seg}.from."
+                                )
+
+            # Check zoning systems are compatible.
+            if self.zoning_system != target.data.zoning_system:
+                zoning_diff = True
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1325]: `zoning_diff` is set but not used anywhere should this be `target.zoning_diff`?

File: src/caf/core/data_structures.py [hunk 9]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+                    )
+                if target.segment_translations is not None:
+                    for targ_seg, seed_seg in target.segment_translations.items():
+                        factor = factor.translate_segment(
+                            from_seg=targ_seg,
+                            to_seg=seed_seg,
+                            reverse=True,
+                            _bypass_validation=bypass,
+                        )
+                new_dvec = new_dvec.__mul__(factor, _bypass_validation=bypass)
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1413]: Method for factoring to targets all looks good, although there is lots of translation happening which keeps creating new DVectors and copying data is this slowing things down at all? In future is it worth looking into any ways to potentially reduce the amount of translations required or potentially performing some calculations outside the loop which can help with them, if speed isn't a problem then not worth bothering but possibly an area to revisit in future if speed is an issue

File: src/caf/core/data_structures.py [hunk 10]
```diff
@@ -706,5 +1448,292 @@ def old_to_new_dvec(import_data: dict):
...
+        # Segment by LAD segment total reports - 1 to 1, No weighting
+        try:
+            lad = ZoningSystem.get_zoning("lad_2020")
+            dvec = self.aggregate(lad_report_seg)
+            dvec = dvec.translate_zoning(lad)
+            dvec.data.to_csv(lad_report_path)
+        except Exception as err:
+            LOG.error("Error creating LAD report: %s", err)
+
+    def sum(self):
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1564]: Type annotations, I think this returns a float?
```suggestion
    def sum(self) -> float:
```

File: src/caf/core/data_structures.py [hunk 11]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+            import_data=new_data,
+            segmentation=new_seg,
+            zoning_system=self.zoning_system,
+            time_format=self.time_format,
+            low_memory=self.low_memory,
+            val_col=self.val_col,
+            cut_read=True,
         )
 
+    def calc_rmse(self, targets: list[IpfTarget]):
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1235]: Return type annotation, I think this returns a single float?
```suggestion
    def calc_rmse(self, targets: list[IpfTarget]) -> float:
```

File: src/caf/core/data_structures.py [hunk 12]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+        Calculate the rmse relative to a set of targets.
+
+        Parameters
+        ----------
+        targets: list[IpfTarget]
+            The targets to calc rmse relative to.
+        """
+        mse = 0
+        for target in targets:
+            check = self.copy()
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1246]: Is this copy strictly necessary?

File: src/caf/core/data_structures.py [hunk 13]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+                        factor = factor.translate_segment(
+                            from_seg=targ_seg,
+                            to_seg=seed_seg,
+                            reverse=True,
+                            _bypass_validation=bypass,
+                        )
+                new_dvec = new_dvec.__mul__(factor, _bypass_validation=bypass)
+
+            rmse = new_dvec.calc_rmse(targets)
+            print(f"RMSE = {rmse} after {i + 1} iterations.")
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1416]: Use `LOG.x` instead of print

File: src/caf/core/data_structures.py [hunk 14]
```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
...
+            if rmse < tol:
+                print("Convergence met, returning DVector.")
+                return new_dvec
+            if abs(rmse - prev_rmse) < tol:
+                print(f"RMSE has stopped improving at {rmse}.")
+                return new_dvec
+            prev_rmse = rmse
+        warnings.warn(
+            "Convergence has not been met, and RMSE has not stopped improving. "
```
Comment (MattBuckley-TfN) [inline src/caf/core/data_structures.py:1425]: Isn't this warning reached when max iterations is met not when RMSE has stopped improving

**Review by MattBuckley-TfN — COMMENTED** [review]
I think the method for IPF looks like its working fine, the only potential error is `validate_ipf_targets` isn't setting `zoning_diff` correctly?

I think it might be worth adding some extra docs and type annotations if someone else needs to use this. Use `collections.abc.Collection` instead of list for parameter type annotations as it is less restrictive, still use specific types (list, tuple ...) for return types.

Also, the main IPF loop seems to be doing a lot of translations and copying data to recreate DVectors if the process is slow it might be worth revisiting this area to try speed things up.

File: src/caf/core/data_structures.py (hunk 9 above)
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:1413]: I do think this may be a problem but at the moment I think all translations are necessary. Possibly the final translations of segment/zoning could be removed (getting factors back to seed target/zoning) if the seed segment(s) and zoning were kept in the DVector when throughout. For segments this would be relatively easy, but for zoning this would mean a refactor to allow multiindexed columns and multiple zone systems in a DVector. I have refactored so that after the first iteration there is no validation on DVector creation which should speed it up a bit.

File: src/caf/core/data_structures.py (hunk 12 above)
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:1246]: Check needs to be created somewhere so I don't think it matters - this just means for all of the possible translations I don't need to add checks for if 'check' already exists.

File: src/caf/core/data_structures.py (hunk 14 above)
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:1425]: Yes, that's why it says RMSE has not stopped improving

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Methodology all looks good, just some style-like suggestions. I've also added some issues for any other changes.

- `tests\segments.xlsx` does this Excel file need to be stored on GitHub? Could the checks it does just be unit / integration tests?
- Run isort and black on repo
- DVectors should defaults to .dvec when saving

File: pyproject.toml [hunk 15]
```diff
@@ -107,7 +108,7 @@ add-ignore = [
 ]
 
 [tool.pytest.ini_options]
-addopts = "--cov=caf.core --cov-report=xml --doctest-modules -n auto"
+#addopts = "--cov=caf.core --cov-report=xml --doctest-modules -n auto"
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply pyproject.toml:111]: This needs to be commented out for debugging to work in tests - I usually don't commit this change but must have by accident

File: tests/segments.csv [hunk 16]
```diff
@@ -0,0 +1,98 @@
+Adding segment,to segment_1,to segment_2
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply tests/segments.csv]: Discussed, will delete

File: src/caf/core/data_structures.py [hunk 17]
```diff
@@ -518,130 +599,384 @@ def translate_zoning(
...
+        trans: pd.DataFrame | None = None
+            The translation to use to split the DVector. If left as None this will be
+            found using the DVector method.
+
+        Returns
+        -------
+        dict[int, DVector]
+            A dict of self split into DVectors for each zone in agg_zoning
+        """
+        if trans is None:
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:696]: I don't think it's currently checked for, but it wouldn't work

File: src/caf/core/data_structures.py [hunk 18]
```diff
@@ -684,8 +1024,636 @@ def aggregate(self, segs: list[str]):
...
+        """
+        target_sum = 0.0
+        for position, target in enumerate(targets):
+            # Check targets sum to the same, or they can't converge. Potentially could allow
+            # IPF for non-agreeing targets to get as close as possible.
+            if target_sum == 0:
+                target_sum = target.data.sum()
+            else:
+                # TODO don't hard code this
+                if not math.isclose(target_sum, target.data.sum(), abs_tol=target_sum / 1e5):
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py]: I think I rushed this for it to work, as the TODO says this probably shouldn't be hard coded and I'll change it now

File: src/caf/core/data_structures.py [hunk 19]
```diff
@@ -684,8 +1024,636 @@ def aggregate(self, segs: list[str]):
...
+                        )
+                new_dvec = new_dvec.__mul__(factor, _bypass_validation=bypass, how="outer")
+
+            rmse = new_dvec.calc_rmse(targets)
+            LOG.info(f"RMSE = {rmse} after {i + 1} iterations.")
+            if rmse < tol:
+                print("Convergence met, returning DVector.")
+                return new_dvec, rmse
+            if abs(rmse - prev_rmse) < tol:
+                LOG.info(f"RMSE has stopped improving at {rmse}.")
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:1660]: This happens so frequently it seems more like expected behaviour. Targets need to be extremely compatible to get to RMSEs users might ask for from experience of furnessing for example.

File: src/caf/core/data_structures.py (hunk 14 above)
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:1664]: Yes that's what this says. It hasn't converged, but RMSE is still getting better by iteration, otherwise either one of the other to conditions would have been met.

File: src/caf/core/segments.py [hunk 20]
```diff
@@ -114,91 +263,198 @@ def get_segment(self, subset: Optional[list[int]] = None):
...
 
+class SegConverter(enum.Enum):
+    """Enum class for converting segment combos."""
+
+    AG_G = "ag_g"
+    APOPEMP_AWS = "apopemp_aws"
+    CARADULT_HHTYPE = "caradult_hhtype"
+    NSSEC_ADULT = "nssec_adult"
+
+    def get_conversion(self):
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/segments.py:289]: Yes agree, but I'll leave it for now

**Review by isaac-tfn — COMMENTED** [review]
Addressed some comments, commented where appropriate. Hopefully can be merged end of this week

File: src/caf/core/data_structures.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/core/data_structures.py:696]: Will it error at some point or will it continue and just produce incorrect results?

[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:696]: I'm being lazy I'll just add an Raise it's only a line or two of code

**Review by MattBuckley-TfN — APPROVED** [review]
I think everything looks good, I think the only remaining comments are around tidying things up and I've made most into issues so they can be dealt with later.

**Comment by MattBuckley-TfN** [comment]
Just noticed it fails on 3.9 because of type annotations, do we need it to work on 3.9?

**Review by MattBuckley-TfN — APPROVED** [review]
